lib/Dialect/StableHLO/Transforms/UpdateGlobalToLocalShapes.cpp

```diff
@@ -16,6 +16,93 @@ namespace mlir::tt::stablehlo {
+                                 mlir::sdy::MeshOp &globalMeshOp) {
+  mlir::stablehlo::ScatterDimensionNumbersAttr scatterDimensionNumbers =
+      scatterOp.getScatterDimensionNumbers();
+  llvm::ArrayRef<int64_t> insertedWindowDims =
+      scatterDimensionNumbers.getInsertedWindowDims();
+  llvm::ArrayRef<int64_t> scatterDimsToOperandDims =
+      scatterDimensionNumbers.getScatterDimsToOperandDims();
+
+  // Get sharding info for inputs
+  llvm::ArrayRef<mlir::sdy::DimensionShardingAttr> inputDimShardings =
+      shardy_utils::getOperandShardingAttr(
+          scatterOp.getOperation()->getOpOperand(0), globalMeshOp)
+          .getDimShardings();
+
+  // Get sharding info for updates
+  llvm::ArrayRef<mlir::sdy::DimensionShardingAttr> updateDimShardings =
+      shardy_utils::getOperandShardingAttr(
+          scatterOp.getOperation()->getOpOperand(2), globalMeshOp)
+          .getDimShardings();
+
+  // Early exit conditions - return false if we can't analyze safely
+  if (inputDimShardings.empty() || updateDimShardings.empty()) {
+    return false;
+  }
+
+  mlir::OperandRange scatterInputs = scatterOp.getInputs();
+  mlir::OperandRange scatterUpdates = scatterOp.getUpdates();
+
+  // shlo scatter accepts a variadic number of inputs / updates tensors, we
+  // expect only 1
+  if (scatterInputs.size() != 1 || scatterUpdates.size() != 1) {
+    return false;
+  }
+
+  // expect the inputs and updates to have the same size
+  mlir::RankedTensorType scatterInputType =
+      mlir::dyn_cast<mlir::RankedTensorType>(scatterInputs.front().getType());
+  mlir::RankedTensorType scatterUpdateType =
+      mlir::dyn_cast<mlir::RankedTensorType>(scatterUpdates.front().getType());
+  if (scatterInputType.getShape().size() !=
+      scatterUpdateType.getShape().size()) {
+    return false;
+  }
+
+  // insertedWindowDims / scatterDimsToOperandDims must be a single dim and
+  // equal for a cache update
+  if (insertedWindowDims.size() != 1 || scatterDimsToOperandDims.size() != 1 ||
+      (insertedWindowDims.front() != scatterDimsToOperandDims.front())) {
+    return false;
+  }
+
+  int64_t scatterAxis = insertedWindowDims.front();
+
+  // figure out sharding axis
+  // input and updates must be sharded equivalently
+  if (!inputDimShardings.equals(updateDimShardings)) {
+    return false;
+  }
+
+  llvm::SmallVector<uint8_t> shardingAxes = {};
+
+  for (auto [dimIndex, sharding] : llvm::enumerate(inputDimShardings)) {
+    if (!sharding.getAxes().empty()) {
+      shardingAxes.push_back(dimIndex);
+    }
+  }
+
+  // We expect the sharding axes and scatter axes to be disjoint
+  if (llvm::is_contained(shardingAxes, scatterAxis)) {
+    return false; // Not safe - scatter axis is sharded
+  }
+
+  scatterOp.emitWarning("In scatterOp ")
+      << scatterOp->getName() << ", the scatter axis " << scatterAxis
+      << " is orthogonal to input sharding axes - this sharded scatter "
+         "is safe to lower through UpdateGlobalToLocalShapesPass. \n";
```

Comment:
No need of newline character for these type of messages.
```suggestion
         "is safe to lower through UpdateGlobalToLocalShapesPass.";
```